### PR TITLE
[libyui-qt-pkg] Fix Pattern Categories Order + Debug Views for Patterns

### PR DESCRIPTION
**Pull request by shundhammer**
## Bugzilla

https://bugzilla.suse.com/show_bug.cgi?id=1216093

## Trello

https://trello.com/c/AKKngvcq

## Problem

During development of a pattern selector for Agama / ALP, @lslezak noticed that the pattern categories had a different order than the ones in the YaST Qt pattern selector.

So we added some debug views (see below) for the patterns to the YaST Qt pattern selector to investigate which one was right, and it turned out that the new one for Agama had the correct order.

patterns-wrong-order

Notice how pattern category "_Graphical Environments_" uses order _1802_ from the _last_ of its patterns "_A very basic desktop_", not _1000_ from "_GNOME Desktop Environment (Basic)_" as it should.

Similar for pattern category "_Desktop Functions_" which uses _2000_ from "_Technical Writing_" instead of _1580_ from "_Multimedia_".

Collapsed pattern categories:

patterns-wrong-order-categories

## Cause

Our software patterns have a field _order_ (type string) to sort them logically rather than alphabetically by name. Each pattern _category_ should pick the first of its patterns and use the _order_ field from that one. But it turned out that the last one was picked instead which was wrong.

## Fix

Now using the correct comparison function to really pick the first pattern and not the last. In addition to that, now handling patterns correctly where that field was omitted in the definition and sorting those patterns last.

patterns-fixed-order

Notice how each pattern category now really uses its _first_ pattern for the order, not the last.

patterns-fixed-order-categories

Collapsed pattern categories in the order that is correct according to the definition. Compare with the other screenshot above with the collapsed categories.

## Better keep it Broken for Consistency with Older Releases?

Should we decide that we don't want the fixed categories order, the new debug views are consistent and self-sufficient commits that can be kept regardless.

Alternatively, the pattern definitions (the .spec files of the pattern packages) could be changed to reflect the desired order.

## What About NCurses?

The NCurses package selector has a patterns filter, but it doesn't show pattern categories.

ncurses-patterns

# New Debug Views

_Available in the Qt UI only!_

## Showing the Pattern _Order_ in an extra Column in the List

(On request, when the new `Y2_SHOW_PATTERNS_ORDER `environment variable is set)

fixed-pattern-category-order

```
sudo Y2_SHOW_PATTERNS_ORDER=1 yast2 sw_single
```

## Showing Invisible Patterns

(On request, when the new `Y2_SHOW_INVISIBLE_PATTERNS` environment variable is set)

invisible-patterns

```
sudo Y2_SHOW_INVISIBLE_PATTERNS=1 yast2 sw_single
```

This is useful for to figure out how a package gets on the system: The user-visible patterns typically consist of a number of _invisible_ patterns as building blocks that often in turn consist of other invisible patterns.

Patterns are just lightweight RPMs, and they use RPM dependencies like _Recommends_ and _Supplements_ to get other patterns (or packages) into the pool of packages to install. But without seeing those patterns directly, it's very inconvenient to piece that dependency tree together to get a picture how and why packages get installed.

With this new debug view, it's considerably easier. As shown in the screenshot, once the pattern is selected, click on its _patterns-xxx_ package in the right ("_Packages_") table and switch to the "_Dependencies_" tab on the bottom panel.

## Using Both at the Same Time

```
sudo Y2_SHOW_PATTERNS_ORDER=1 Y2_SHOW_INVISIBLE_PATTERNS=1 yast2 sw_single
```

### CODE REVIEW

File: libyui-qt-pkg/src/YQPkgPatternList.cc [hunk 1]
```diff
@@ -361,6 +376,19 @@ YQPkgPatternListItem::init()
...
 	setIcon( _patternList->iconCol(), YQUI::ui()->loadIcon( iconName ) );
 
+        if ( _patternList->showInvisiblePatterns() && ! _zyppPattern->userVisible() )
+        {
+            // The YQPkgObjListItem base class takes care of setting the name
+            // and summary columns, but here we want to add something to it.
+            // Notice that patterns use the summary column.
+
+            QString name = text( _patternList->summaryCol() );
+            name += " [Invisible]";
```
Comment (lslezak) [inline libyui-qt-pkg/src/YQPkgPatternList.cc:386]: Although this is used only in a special debugging case maybe we should still mark it for translation... :thinking:

Comment (shundhammer) [reply libyui-qt-pkg/src/YQPkgPatternList.cc:386]: I am not so sure about that... it's a very special environment.

Comment (shundhammer) [reply libyui-qt-pkg/src/YQPkgPatternList.cc]: Against my better judgement...
But I had to force-push to keep the history clean and understandable.

**Review by lslezak — APPROVED** [review]
LGTM
Thanks for debugging and fixing! :+1:

**Comment by shundhammer** [comment]
Submitted manually to Factory: https://build.opensuse.org/request/show/1116919

**Pull request merged**